### PR TITLE
Convenience methods for producing run input files

### PR DESCRIPTION
**Pull request by Yoshanuikabundi**
### Description

Using the `MDConfig` object to produce run config files is an extra step that this PR eliminates by providing additional export methods on `Interchange`. This PR also tidies up generated `.MDP` files to make them easier for users to modify.

- [x] Add `Interchange.to_mdp()`
- [x] Add `Interchange.to_lammps_input()`
- [x] Add `Interchange.to_sander_input()`
- [x] Make `Interchange.to_gromacs()` produce a `.mdp` file in addition to `.gro` and `.top` files
- [x] Add `Interchange.to_lammps_datafile()` and modify `Interchange.to_lammps()` to produce both input and data files
    - I can see this being controversial - I'm happy to reverse it, but the way I've written it should avoid breaking changes. The `to_lammps(file_path)` argument can be renamed `prefix` and the extension check removed in a later breaking change. The fact that LAMMPS input files include the name of the data file means that producing both files in one step makes a lot of sense - this way users don't have to repeat the name of the data file in subsequent method calls.
- [x] Make the data file referenced by the LAMMPS input file configurable
- [ ] Add `Interchange.to_amber()` to produce all Amber files
    - I haven't done this because Amber has two MD engines, SANDER and PMEMD, and it seems odd to only produce a run input file for the old one when a user asks for Amber files.
- [x] Write and update a lot of relevant documentation
- [x] Neaten up MDP files produced by Interchange so that users can read them and figure out which bits they can modify more easily
- [x] Build up MD config files in memory and only write to disk if no errors are encountered
     - This avoids a situation in which a call to create a config file has failed and so the file is only partially (over)written

### Checklist

- [ ] Add tests
- [x] Lint
- [x] Update docstrings

### CODE REVIEW

**Comment by Yoshanuikabundi** [comment]
I ran into test failures because several tests created Interchanges without boxes, and then used `to_gromacs` to export them. This was previously successful, as .gro and .top files don't require boxes, but now fails because vacuum simulations are not supported by modern versions of GROMACS.

I've corrected the issue by adding boxes to the tests, but this might be considered a breaking change. I can think of a few solutions if this is problematic:

1. Add support for the `Coulomb` electrostatics method to `MDConfig.write_mdp_file()`, even though the generated MDP files will not run on modern versions of GROMACS
2. Skip writing MDP files in `Interchange.to_gromacs()` when it raises an error (and remove this special case in a future breaking change)
3. Revert the changes to `Interchange.to_gromacs()`
4. Do nothing

Let me know what you'd like to do!

**Comment by mattwthompson** [comment]
@timbernat you may like the LAMMPS changes here - if you have some bandwidth, could you have a look?

**Comment by mattwthompson** [comment]
@Yoshanuikabundi

* I'd prefer these changes to point to the `develop` branch and land in 0.4. The downside is this adds ~6 weeks to them going live, but the upside is more freedom to change behavior. What do you think?
* Your potentially-controversial LAMMPS changes are a net positive, I think
* I'm moderately in favor of having `to_amber` write out a `sander` input file analogously to how these changes write out LAMMPS and GROMACS files. Your point about Amber having multiple engines is valid, but
  * It's nice to prepare a simulation with 1 LOC instead of 2-3
  * I think I can only access `sander` so I'm okay focusing on it for now (it'd be nice to get some organizational clarity on this @j-wags)
  * The limitation could be documented
  * I optimistically think the run file will be ignored most of the time and not confuse too many people
  * I expect power-users to ignore the files they don't need, anyway
* I'm torn on possibly changing the behavior of `.to_gromacs` to error out for vacuum systems. You laying out some options is really helpful to me. I've thought about this for a bit now and have some opinions forming, but before I commit to anything I'd like to ask you - do you have opinions on which of the four are particularly good/bad/best?

**Comment by mattwthompson** [comment]
Thanks for the feedback -

* I don't feel strongly at all about how the suffix and stem are extracted, to me that's an implementation detail I can take any way that work
* I prefer using `.lmp` for the sake of simplicity and continuity. You're right people use all sorts of extensions (I inherited `in.simulation_name` scripts when I started grad school, which confused me endlessly until I learned how arbitrary most file names are!) but this sort of thing has to be pushed onto the users' responsibility for practical reasons, I think

**Comment by timbernat** [comment]
That division of responsibility makes a lot of sense, maybe then just warning the user if they make a non-canonical choice of extension? Perhaps replacing [558-562](https://github.com/openforcefield/openff-interchange/pull/1027/commits/acd5efbb01b0d8729e84e26efc2c08ffeef4f611#diff-a3087103ffe505eaa0ffb73f557bffdae70a0862de3d0b802a8207167c09814fR558-R562) with:
```
datafile_path = Path(datafile_path) # path-ify to take guesswork out of name/extension separation
if datafile_path.suffix != '.lmp':
    datafile_path = datafile_path.with_suffix('.lmp')
    warnings.warn(f'Auto-converted extension from "{datafile_path.suffix}" to ".lmp"'

datafile_path = str(datafile_path) # re-stringify to guarantee support with downstream writers
```

**Comment by Yoshanuikabundi** [comment]
@mattwthompson 

> I'd prefer these changes to point to the develop branch and land in 0.4. The downside is this adds ~6 weeks to them going live, but the upside is more freedom to change behavior. What do you think?

Yeah I think that's a good plan. That lets us come up with some better solutions regarding @timbernat's concerns with `to_lammps()`. I'll rebase and change the target branch!

> I'm moderately in favor of having to_amber write out a sander input file analogously to how these changes write out LAMMPS and GROMACS files. 

What about instead of `to_amber()` we call it `to_sander()`?

> I'm torn on possibly changing the behavior of .to_gromacs to error out for vacuum systems. You laying out some options is really helpful to me. I've thought about this for a bit now and have some opinions forming, but before I commit to anything I'd like to ask you - do you have opinions on which of the four are particularly good/bad/best?

I think (2) (skip failing MDP files) is a bad idea because elsewhere we document and commit to the idea that if we can't export something we will error out. I think (1) (add support for nonperiodic electrostatics) might weirdly be the best idea:

- Old versions of GROMACS that support non-periodic systems are still used a lot
- The actual MDP option that is no longer supported is (GROMACS docs are down right now but IIRC) `pbc=no`, which we already write out
- Future GROMACS versions may bring nonperiodic systems back - IIRC the main stated reason for dropping `pbc=no` was that no one was maintaining it
- We can include a comment in the MDP file describing why the MDP file isn't working

@timbernat 
> In [to_lammps_input()](https://github.com/openforcefield/openff-interchange/pull/1027/commits/acd5efbb01b0d8729e84e26efc2c08ffeef4f611#diff-a3087103ffe505eaa0ffb73f557bffdae70a0862de3d0b802a8207167c09814fR578-R605): this is great!, swapping out the input file path is a feature I've been long awaiting, and I appreciate no longer needing to import MDConfig for surface-level writes :P

Yay!

> > In [to_lammps()](https://github.com/openforcefield/openff-interchange/pull/1027/commits/acd5efbb01b0d8729e84e26efc2c08ffeef4f611#diff-a3087103ffe505eaa0ffb73f557bffdae70a0862de3d0b802a8207167c09814fR557-R560): why use stringification to extract prefix? Would be cleaner to use Path.suffix and Path.name attrs IMHO (you actually do a variant of this [below](https://github.com/openforcefield/openff-interchange/pull/1027/commits/acd5efbb01b0d8729e84e26efc2c08ffeef4f611#diff-a3087103ffe505eaa0ffb73f557bffdae70a0862de3d0b802a8207167c09814fR600-R601)). Also, I've seen different extensions used for the input file besides .lmp (e.g. I and others use .lammps, and I've even seen some reverse the order of filename and extension viz. lmp.compound rather than compound.lmp) . Hard-coding in the .lmp suffix might be a little dangerous for extensibility and lead to unexpected behavior for inexperienced user, who may not notice your [explicit documentation](https://github.com/openforcefield/openff-interchange/pull/1027/commits/9e8d6bcd668dc90c9f5e719effdf523e20ff1a69#diff-a3087103ffe505eaa0ffb73f557bffdae70a0862de3d0b802a8207167c09814fR551-R554) at first glance.

Thanks for this feedback, that's really helpful context!! I think if we're targeting 0.4.0, we can side step the issue by renaming the `file_path` argument to `prefix` and dropping the handling of the extension, which would be less surprising and more consistent with the existing `to_gromacs` API. 

I could also add optional `datafile_path` and `inputfile_path` arguments that users can use to specify their own file names? In `to_gromacs()` this is unnecessary because users might as well just call `to_gro()`, `to_top()`, and `to_mdp()`, but for LAMMPS this would mean repeating the data file name in calls to both `to_lammps_datafile()` and `to_lammps_inputfile` so there's a strong ergonomics argument here.

**Comment by mattwthompson** [comment]
I don't completely agree with your conclusions/preferences in ways that might keep this thread going for longer than it needs to. IIUC, in its current state, this PR does what you set out to do % 
* a `to_amber` method
* how best to handle non-periodic systems written out to GROMACS

Since we're targeting a release with some behavior changes (that might not come out for a while), I'm just going to review this as-is and we can hash out the remaining two points later.

On the LAMMPS bit

> we can side step the issue by renaming ... and dropping the handling of the extension

+1 from me

File: openff/interchange/components/mdconfig.py [hunk 1]
```diff
@@ -18,13 +20,15 @@
...
-nsteps                   = 0
-nstenergy                = 1000
-continuation             = yes
-cutoff-scheme            = verlet
+MDP_HEADER = """\
+; Single point energy
+nsteps                   = 0      ; Perform no MD steps
+nstenergy                = 1000   ; Write energies to .EDR file
+continuation             = yes    ; Do not apply constraints at start
+cutoff-scheme            = verlet ; Use a Verlet pair list for cutoffs
```
Comment (mattwthompson) [inline openff/interchange/components/mdconfig.py:28]: Not worth doing here, but this file screams out to be stored in memory with a little class that inherits from `NamedTuple` or something

**Review by mattwthompson — APPROVED** [review]
Great work, I like the features added and the user-facing simplicity of the added methods

`Interchange.to_gromacs` _failing_ when a box is not specified is a behavior change, but we'll hash that out elsewhere

File: docs/using/output.md [hunk 2]
```diff
@@ -1,98 +1,62 @@
 # Exporting to other software
 
-`Interchange` provides several methods to produce input data for other
-software. Note that none of these methods write out all the information
-stored in an `Interchange`; they support a design where the principle
-source of truth is the rich chemical information in the `Interchange`
-object, and exported files are tools to perform some operation.
+`Interchange` provides several methods to produce input data for other software. Note that none of these methods write out all the information stored in an `Interchange`; they support a design where the principle source of truth is the rich chemical information in the `Interchange` object, and exported files are tools to perform some operation.
```
[the remark this answers is not in the document]
Comment (Yoshanuikabundi) [reply docs/using/output.md:3]: Yeah I think I've vacillated on this a bit but a year or so ago I settled on soft wrapping within paragraphs in prose as it makes things much easier to read when you don't have <wrap length> columns of text in the viewer. I'll have a squiz for a tool!

File: docs/using/output.md [hunk 3]
```diff
@@ -1,98 +1,62 @@
...
-An input file can be written from an [MDConfig object] constructed from the interchange. The resulting file will run a single-point energy calculation and
-should be modified for the desired simulation:
-
-```python
-mdconfig.write_lammps_input(input_file="auto_generated.in")
-```
+Note that the generated run input file will run a single-point energy calculation and should be modified for the desired simulation.
 
-Note that LAMMPS does not seem to implement a switching function as [commonly used](https://openforcefield.github.io/standards/standards/smirnoff/#vdw) by SMIRNOFF force fields.
+LAMMPS does not implement a switching function as [commonly used](https://openforcefield.github.io/standards/standards/smirnoff/#vdw) by SMIRNOFF force fields, so these force fields will produce different results in LAMMPS than in OpenMM or GROMACS.
```
[the remark this answers is not in the document]
Comment (Yoshanuikabundi) [reply docs/using/output.md:47]: I figure the fastest way to be corrected on the internet is to confidently say something wrong!

File: openff/interchange/_tests/unit_tests/components/test_mdconfig.py [hunk 4]
```diff
@@ -44,6 +44,8 @@ def parse_mdp(file: str) -> dict[str, str]:
 
     with open(file) as f:
         for line in f.readlines():
+            line, _semicolon, _comment = line.partition(";")
+
```
[the remark this answers is not in the document]
Comment (Yoshanuikabundi) [reply openff/interchange/_tests/unit_tests/components/test_mdconfig.py:48]: Correct.

File: openff/interchange/components/mdconfig.py (hunk 1 above)
Comment (Yoshanuikabundi) [reply openff/interchange/components/mdconfig.py:28]: Mmm I was thinking about something like that

**Comment by mattwthompson** [comment]
Alright - I think the next steps here are to get tests running (I thought I set up the workflow to run against `develop` ...) and merge when everything is green. Then the leftovers from [this comment](https://github.com/openforcefield/openff-interchange/pull/1027#issuecomment-2278551779) will be handled downstream

**Pull request merged**